Improve domjudge infrastructure

incaseoftrouble opened the pull request:
Closes #94 

Sorry for throwing multiple things into one, I can pick them apart if needed.

Things I added:
 * php version is now automatically detected
 * `make dist` is run for the `build.sh` so one can use a local tarball
 * "one-shot" customization done by `start.sh` (e.g. php configuration) are moved into the dockerfile (`configure.sh`), so one could overwrite these changes with a bind mount
 * `start.sh` now runs additional scripts at the end (can be supplied with bind mounts)
 * small improvements to the dockerfile

The `start.sh` infrastructure actually is a bit like init.d so one could even break apart the start.sh script for ease of maintenance.

nickygerritsen reviewed (approved):
I like it. Let’s see if someone else finds something otherwise I will merge.

It’s fine all is in one.

Perhaps update the readme? I can then update the docker hub page based on it

incaseoftrouble commented:
I'll do that!

nickygerritsen reviewed (approved):
New changes look good too

incaseoftrouble commented:
One more change: After trying it out, I found out the following problem:
 * During the start script, the symfony cache is created
 * The "add remote user" scripts modify the firewall settings
 * `webapp cache:clear` generates a new container (with the new settings)
 * BUT also leaves the old container in the `var/cache/prod/` directory - so there are two `ContainerXXXXX` folders there
 * They are named randomly and the first one is chosen
 * => The new settings are not applied on some starts

This can be fixed by running `webapp cache:clear` twice or applying the modifications before the cache is first generated. I thus split up the `start.sh` script in parts, so that one can (somewhat) choose when the modifications are applied.

incaseoftrouble commented:
Ok, everything should be fixed now. I tested the build locally.

I also modified the judgehost builds to be "symmetric" in structure as far as possible.

I also split up the overall build script a bit, so you can `./build-domjudge my_local_domjudge`, useful if you work on a fork.

I only need to figure out how to make it build on gitlab - but thats some subtle issue with the default shell there ...

mvr320 commented on .gitlab-ci.yml:
```diff
@@ -121,7 +121,7 @@ check-pr-DOMjudge:
   script:
     - HUBURL="https://registry.hub.docker.com/v1/repositories/domjudge/domserver/tags"
-    - apk add jq curl
+    - apk add jq curl bash
```
if you keep the `sh ./build` I think adding bash is not strictly needed.

incaseoftrouble replied:
But then you get a weirdness on Debian / Ubuntu-esque systems: `sh` may not have `[[` available and the build fails there. (The build.sh script would call `sh build-judgehost.sh`, ignoring the shebang there).

I can of course also revert the change to `build.sh`, I just think it might be useful.

mvr320 replied:
Makes sense.

eldering replied:
I've not followed the conversation in detail, but looking at these scripts, is there any reason for making them bash scripts instead of pure posix shell? All the `[[ -f foobar ]]` can simply be replaced by `[ -f foobar ]`.

incaseoftrouble replied:
Hmmm, but `set -euxo pipefail` is not POSIX, right? (I know very little about what is POSIX though).

eldering replied:
`set -eux` does exist in posix, see https://man7.org/linux/man-pages/man1/set.1p.html

incaseoftrouble replied:
Ah, I'll adapt accordingly

incaseoftrouble replied:
Ok, it seems to work now both locally and on the CI and I think it should all be posix.

eldering commented on docker/domserver/configure.sh:
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash -e
```
Is it possible to also use `/bin/sh` here, as well as in the scripts below?

If there's a really good reason not to, that's ok, but I think it's better to try and use posix shell.

incaseoftrouble replied:
I think it should work, I just used bash by habit. I'll have a look later. Thanks for you feedback.

incaseoftrouble replied:
It works everywhere except in the domjudge initialization which requires indirect substitution for the `file_or_env` thing

eldering commented on docker/judgehost/configure.sh:
```diff
@@ -0,0 +1,11 @@
+chown -R domjudge: /opt/domjudge
+
+chmod 755 /scripts/start.sh
+if compgen -G "/scripts/bin/*"
```
POSIX shell can be used here by looping and checking that each file exists:
```sh
for i in /scripts/bin/* ; do
    if [ -f "$i" ]; then
        ...
    fi
done
```

mvr320 commented:
@incaseoftrouble do you want to fixup the last commit or can it be merged? We'll push new containers after its merged.

incaseoftrouble commented:
I can fixup if you prefer but I guess the posix change is a sensible commit by itself :)
